[MPS] Add Metal implementation of exp op

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128421
* #128375
* #128373

To improve accuracy, use `precise::exp()` (and `precise::sin()`/`precise::cos()` for complex flavor)
Reuse `test_exp1` to check that accuracy of `exp` ops is sometimes closer to CPU

Fix bug in non-contiguous tensors handling

Fixes https://github.com/pytorch/pytorch/issues/84936

malfet committed:
[MPS] Add Metal implementation of exp op

To improve accuracy, use only `exp` from `precise::` namespace

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add Metal implementation of exp op"


To improve accuracy, use`precise::exp()` 

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add Metal implementation of exp op"


To improve accuracy, use`precise::exp()` 

[ghstack-poisoned]

kulinseth commented on aten/src/ATen/native/mps/UnaryConstants.h:
```diff
@@ -35,4 +35,18 @@ kernel void erfinv_mps_kernel( device {0} *output [[buffer(0)]],
-}})METAL";
+}}
+
+kernel void exp_kernel( device {0} *output [[buffer(0)]],
```
Can you add tanh as well ?

malfet replied:
Sure, added, those complex implementation gets a bit hairy...

kulinseth commented:
@malfet , can we please add a test to make sure exp and tanh are precise.

malfet committed:
Update on "[MPS] Add Metal implementation of exp op"


To improve accuracy, use `precise::exp()` (and `precise::sin()`/`precise::cos()` for complex flavor)

Fix bug in non-contiguous tensors handling

[ghstack-poisoned]

malfet commented:
Can we please add a test to make sure exp and tanh are precise.

@kulinseth , do you have some examples in mind? I.e. see this comment https://github.com/pytorch/pytorch/issues/84936#issuecomment-2142220367 exp on CPU and MPS do not match, even if one uses high-precision variants

malfet committed:
Update on "[MPS] Add Metal implementation of exp op"


To improve accuracy, use `precise::exp()` (and `precise::sin()`/`precise::cos()` for complex flavor)

Fix bug in non-contiguous tensors handling

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Add Metal implementation of exp op"


To improve accuracy, use `precise::exp()` (and `precise::sin()`/`precise::cos()` for complex flavor)

Fix bug in non-contiguous tensors handling

Fixes https://github.com/pytorch/pytorch/issues/84936 

[ghstack-poisoned]

kulinseth reviewed (approved):
Looks good.
